File: src/leadforge/api/routes/grants.py
import uuid
from datetime import datetime, timezone

import structlog
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from leadforge.api.deps import get_db
from leadforge.api.schemas.grant import (
    GrantApplicationCreate,
    GrantApplicationResponse,
    GrantApplicationUpdate,
    GrantBoardCard,
    GrantBoardColumn,
    GrantDocumentResponse,
    GrantDocumentUpdate,
    GrantFinancialsResponse,
    GrantStageTransition,
)
from leadforge.db.models.business import Business
from leadforge.db.models.grant_application import GrantApplication, NOFStage
from leadforge.db.models.grant_document import GrantDocument
from leadforge.grants.financial_calculator import compute_grant_financials

logger = structlog.get_logger()

router = APIRouter(prefix="/grants", tags=["grants"])

# Valid stage transitions — enforced by backend
VALID_NOF_TRANSITIONS: dict[NOFStage, set[NOFStage]] = {
    NOFStage.ELIGIBILITY_ASSESSED: {NOFStage.INTAKE, NOFStage.REMOVED},
    NOFStage.INTAKE: {NOFStage.APPLIED, NOFStage.REMOVED},
    NOFStage.APPLIED: {NOFStage.PIPELINE, NOFStage.REMOVED},
    NOFStage.PIPELINE: {NOFStage.FINALIST, NOFStage.REMOVED},
    NOFStage.FINALIST: {NOFStage.STAGE_1_LEGAL, NOFStage.REMOVED},
    NOFStage.STAGE_1_LEGAL: {NOFStage.STAGE_2_DOCS, NOFStage.REMOVED},
    NOFStage.STAGE_2_DOCS: {NOFStage.STAGE_3_FINANCING, NOFStage.REMOVED},
    NOFStage.STAGE_3_FINANCING: {NOFStage.STAGE_3_CONSTRUCTION, NOFStage.REMOVED},
    NOFStage.STAGE_3_CONSTRUCTION: {NOFStage.STAGE_4_CLOSING, NOFStage.REMOVED},
    NOFStage.STAGE_4_CLOSING: {NOFStage.STAGE_5_COMPLETE, NOFStage.REMOVED},
    NOFStage.STAGE_5_COMPLETE: {NOFStage.ALUMNUS, NOFStage.REMOVED},
    NOFStage.ALUMNUS: set(),
    NOFStage.REMOVED: set(),
}

# Board groupings
BOARD_GROUPS = {
    "Pre-Application": [
        NOFStage.ELIGIBILITY_ASSESSED,
        NOFStage.INTAKE,
        NOFStage.APPLIED,
        NOFStage.PIPELINE,
    ],
    "Active Grant": [
        NOFStage.FINALIST,
        NOFStage.STAGE_1_LEGAL,
        NOFStage.STAGE_2_DOCS,
        NOFStage.STAGE_3_FINANCING,
        NOFStage.STAGE_3_CONSTRUCTION,
        NOFStage.STAGE_4_CLOSING,
        NOFStage.STAGE_5_COMPLETE,
    ],
    "Complete": [
        NOFStage.ALUMNUS,
        NOFStage.REMOVED,
    ],
}


@router.get("/board")
async def get_grant_board(session: AsyncSession = Depends(get_db)):
    """Get grant Kanban board grouped by stage."""
    # Stage counts
    count_q = select(GrantApplication.status, func.count(GrantApplication.id)).group_by(
        GrantApplication.status
    )
    counts = dict((await session.execute(count_q)).all())

    columns = []
    for group_name, stages in BOARD_GROUPS.items():
        for stage in stages:
            count = counts.get(stage, 0)

            # Fetch up to 10 preview cards per stage
            preview_q = (
                select(GrantApplication, Business.name)
                .join(Business, GrantApplication.business_id == Business.id)
                .where(GrantApplication.status == stage)
                .order_by(GrantApplication.updated_at.desc())
                .limit(10)
            )
            results = (await session.execute(preview_q)).all()

            now = datetime.now(timezone.utc)
            cards = [
                GrantBoardCard(
                    grant_id=str(row[0].id),
                    business_id=str(row[0].business_id),
                    business_name=row[1],
                    corridor_name=row[0].corridor_name,
                    estimated_grant=row[0].base_grant_amount,
                    days_in_stage=(now - row[0].updated_at).days
                    if row[0].updated_at
                    else 0,
                )
                for row in results
            ]

            columns.append(
                GrantBoardColumn(
                    stage=stage.value,
                    count=count,
                    cards=cards,
                )
            )

    return {"columns": [col.model_dump() for col in columns]}


@router.get("/financials/{grant_id}", response_model=GrantFinancialsResponse)
async def get_grant_financials(
    grant_id: uuid.UUID,
    session: AsyncSession = Depends(get_db),
):
    """Compute grant financials for an application."""
    result = await session.execute(
        select(GrantApplication).where(GrantApplication.id == grant_id)
    )
    grant = result.scalar_one_or_none()
    if not grant:
        raise HTTPException(status_code=404, detail="Grant application not found")

    financials = compute_grant_financials(
        total_project_cost=grant.total_project_cost or 0.0,
        acquisition_cost=grant.acquisition_cost or 0.0,
    )
    return GrantFinancialsResponse(
        total_project_cost=financials.total_project_cost,
        acquisition_cost=financials.acquisition_cost,
        base_grant=financials.base_grant,
        taf_eligible=financials.taf_eligible,
        owner_contribution=financials.owner_contribution,
        owner_min_financing=financials.owner_min_financing,
        exterior_work_minimum=financials.exterior_work_minimum,
    )


@router.get("/{grant_id}", response_model=GrantApplicationResponse)
async def get_grant(
    grant_id: uuid.UUID,
    session: AsyncSession = Depends(get_db),
):
    """Get full grant application detail with documents."""
    result = await session.execute(
        select(GrantApplication)
        .options(selectinload(GrantApplication.documents))
        .where(GrantApplication.id == grant_id)
    )
    grant = result.scalar_one_or_none()
    if not grant:
        raise HTTPException(status_code=404, detail="Grant application not found")
    return grant


@router.get("/", response_model=list[GrantApplicationResponse])
async def list_grants(
    status: str | None = Query(None),
    corridor_name: str | None = Query(None),
    business_id: uuid.UUID | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    session: AsyncSession = Depends(get_db),
):
    """List grant applications with filters."""
    stmt = select(GrantApplication)

    if status:
        try:
            stage = NOFStage(status)
        except ValueError:
            raise HTTPException(status_code=400, detail=f"Invalid status: {status}")
        stmt = stmt.where(GrantApplication.status == stage)

    if corridor_name:
        stmt = stmt.where(GrantApplication.corridor_name == corridor_name)

    if business_id:
        stmt = stmt.where(GrantApplication.business_id == business_id)

    stmt = stmt.order_by(GrantApplication.created_at.desc())
    stmt = stmt.offset((page - 1) * page_size).limit(page_size)

    result = await session.execute(stmt)
    grants = result.scalars().all()
    return grants


@router.post("/", response_model=GrantApplicationResponse, status_code=201)
async def create_grant(
    body: GrantApplicationCreate,
    session: AsyncSession = Depends(get_db),
):
    """Create a new grant application."""
    # Verify business exists
    biz_result = await session.execute(
        select(Business).where(Business.id == body.business_id)
    )
    if not biz_result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Business not found")

    grant = GrantApplication(
        business_id=body.business_id,
        status=NOFStage.ELIGIBILITY_ASSESSED,
        total_project_cost=body.total_project_cost,
        acquisition_cost=body.acquisition_cost,
        project_description=body.project_description,
    )
    session.add(grant)
    await session.commit()
    await session.refresh(grant)
    logger.info("grant_application_created", grant_id=str(grant.id))
    return grant


@router.patch("/{grant_id}", response_model=GrantApplicationResponse)
async def update_grant(
    grant_id: uuid.UUID,
    body: GrantApplicationUpdate,
    session: AsyncSession = Depends(get_db),
):
    """Partially update a grant application."""
    result = await session.execute(
        select(GrantApplication).where(GrantApplication.id == grant_id)
    )
    grant = result.scalar_one_or_none()
    if not grant:
        raise HTTPException(status_code=404, detail="Grant application not found")

    update_data = body.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(grant, field, value)

    await session.commit()
    await session.refresh(grant)
    logger.info("grant_application_updated", grant_id=str(grant_id))
    return grant


@router.patch("/{grant_id}/stage")
async def transition_grant_stage(
    grant_id: uuid.UUID,
    body: GrantStageTransition,
    session: AsyncSession = Depends(get_db),
):
    """Transition a grant application to a new stage."""
    result = await session.execute(
        select(GrantApplication).where(GrantApplication.id == grant_id)
    )
    grant = result.scalar_one_or_none()
    if not grant:
        raise HTTPException(status_code=404, detail="Grant application not found")

    try:
        new_stage = NOFStage(body.new_stage)
    except ValueError:
        raise HTTPException(status_code=400, detail=f"Invalid stage: {body.new_stage}")

    current_stage = grant.status
    allowed = VALID_NOF_TRANSITIONS.get(current_stage, set())
    if new_stage not in allowed:
        raise HTTPException(
            status_code=422,
            detail=(
                f"Cannot transition from {current_stage.value} "
                f"to {new_stage.value}. "
                f"Allowed: {[s.value for s in allowed]}"
            ),
        )

    grant.status = new_stage
    await session.commit()
    logger.info(
        "grant_stage_transitioned",
        grant_id=str(grant_id),
        from_stage=current_stage.value,
        to_stage=new_stage.value,
    )
    return {
        "status": "ok",
        "grant_id": str(grant_id),
        "new_stage": new_stage.value,
    }


@router.get("/{grant_id}/documents", response_model=list[GrantDocumentResponse])
async def list_grant_documents(
    grant_id: uuid.UUID,
    session: AsyncSession = Depends(get_db),
):
    """Get document checklist for a grant application."""
    # Verify grant exists
    grant_result = await session.execute(
        select(GrantApplication).where(GrantApplication.id == grant_id)
    )
    if not grant_result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Grant application not found")

    result = await session.execute(
        select(GrantDocument).where(GrantDocument.grant_application_id == grant_id)
    )
    return result.scalars().all()


@router.patch("/{grant_id}/documents/{doc_id}", response_model=GrantDocumentResponse)
async def update_grant_document(
    grant_id: uuid.UUID,
    doc_id: uuid.UUID,
    body: GrantDocumentUpdate,
    session: AsyncSession = Depends(get_db),
):
    """Update a grant document status."""
    result = await session.execute(
        select(GrantDocument).where(
            GrantDocument.id == doc_id,
            GrantDocument.grant_application_id == grant_id,
        )
    )
    doc = result.scalar_one_or_none()
    if not doc:
        raise HTTPException(status_code=404, detail="Grant document not found")

    update_data = body.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(doc, field, value)

    await session.commit()
    await session.refresh(doc)
    logger.info("grant_document_updated", doc_id=str(doc_id), grant_id=str(grant_id))
    return doc

File: api/src/routes/grants.ts
import { Hono } from 'hono';
import { z } from 'zod';
import { requireAuth, requireAdmin } from '../middleware/auth';
import { generateId } from '../lib/jwt';
import { jsonBody, queryParams } from '../lib/validate';
import { computeGrantFinancials } from '../lib/grants';
import { BOARD_GROUPS, NOF_STAGES, VALID_NOF_TRANSITIONS, type NofStage } from '../lib/stages';
import { nowIso, withBooleans, DOCUMENT_BOOLS, GRANT_BOOLS } from '../db/serialize';
import type { AppEnv, GrantApplicationRow, GrantDocumentRow } from '../types';

const router = new Hono<AppEnv>();
const serializeGrant = (row: GrantApplicationRow) => withBooleans<Pick<GrantApplicationRow, keyof GrantApplicationRow>>(row, GRANT_BOOLS);
const serializeDoc = (row: GrantDocumentRow) => withBooleans<Pick<GrantDocumentRow, keyof GrantDocumentRow>>(row, DOCUMENT_BOOLS);

async function loadGrant(db: D1Database, id: string) {
  return db.prepare('SELECT * FROM grant_applications WHERE id = ?').bind(id).first<GrantApplicationRow>();
}

const listQuery = z.object({
  status: z.string().optional(),
  corridor_name: z.string().optional(),
  business_id: z.string().optional(),
  page: z.coerce.number().int().min(1).default(1),
  page_size: z.coerce.number().int().min(1).max(100).default(20),
});

const createBody = z.object({
  business_id: z.string(),
  total_project_cost: z.number().nullable().optional(),
  acquisition_cost: z.number().nullable().optional(),
  project_description: z.string().nullable().optional(),
});

const UPDATE_FIELDS = [
  'total_project_cost', 'base_grant_amount', 'acquisition_cost', 'taf_amount', 'owner_contribution',
  'financing_amount', 'financing_verified', 'gc_bid_amount', 'project_description', 'exterior_work_pct',
  'has_site_control', 'site_control_type', 'assigned_to', 'ta_provider', 'notes',
] as const;
const num = z.number().nullable().optional();
const str = z.string().nullable().optional();
const bool = z.boolean().nullable().optional();
const updateBody = z.object({
  total_project_cost: num, base_grant_amount: num, acquisition_cost: num, taf_amount: num, owner_contribution: num,
  financing_amount: num, financing_verified: bool, gc_bid_amount: num, project_description: str, exterior_work_pct: num,
  has_site_control: bool, site_control_type: str, assigned_to: str, ta_provider: str, notes: str,
});

const docUpdateBody = z.object({ status: str, notes: str, received_date: str, reviewed_date: str });

// =py routes/grants.get_grant_board
router.get('/board', requireAuth, async (c) => {
  const db = c.env.DB;
  const counts = await db.prepare('SELECT status, COUNT(*) AS n FROM grant_applications GROUP BY status').all<{ status: string; n: number }>();
  const countByStage = new Map((counts.results ?? []).map((r) => [r.status, r.n]));

  const cardStmt = db.prepare(`
    SELECT g.id AS grant_id, g.business_id, b.name AS business_name, g.corridor_name,
           g.base_grant_amount AS estimated_grant, g.updated_at
    FROM grant_applications g JOIN businesses b ON b.id = g.business_id
    WHERE g.status = ? ORDER BY g.updated_at DESC LIMIT 10`);
  const results = await db.batch<{ grant_id: string; business_id: string; business_name: string; corridor_name: string | null; estimated_grant: number | null; updated_at: string }>(
    BOARD_GROUPS.map((stage) => cardStmt.bind(stage))
  );
  const now = Date.now();
  const columns = BOARD_GROUPS.map((stage, i) => ({
    stage,
    count: countByStage.get(stage) ?? 0,
    cards: (results[i].results ?? []).map(({ updated_at, ...card }) => ({
      ...card,
      days_in_stage: Math.max(0, Math.floor((now - Date.parse(updated_at)) / 86_400_000)),
    })),
  }));
  return c.json({ columns });
});

// =py routes/grants.get_grant_financials
router.get('/financials/:grant_id', requireAuth, async (c) => {
  const grant = await loadGrant(c.env.DB, c.req.param('grant_id')!);
  if (!grant) return c.json({ detail: 'Grant application not found' }, 404);
  return c.json(computeGrantFinancials(grant.total_project_cost ?? 0, grant.acquisition_cost ?? 0));
});

// =py routes/grants.get_grant
router.get('/:grant_id', requireAuth, async (c) => {
  const grant = await loadGrant(c.env.DB, c.req.param('grant_id')!);
  if (!grant) return c.json({ detail: 'Grant application not found' }, 404);
  return c.json(serializeGrant(grant));
});

// =py routes/grants.list_grants
router.get('/', requireAuth, queryParams(listQuery), async (c) => {
  const q = c.req.valid('query');
  const where: string[] = [];
  const binds: unknown[] = [];
  if (q.status !== undefined) {
    if (!(NOF_STAGES as readonly string[]).includes(q.status)) return c.json({ detail: `Invalid status: ${q.status}` }, 400);
    where.push('status = ?'); binds.push(q.status);
  }
  if (q.corridor_name) { where.push('corridor_name = ?'); binds.push(q.corridor_name); }
  if (q.business_id) { where.push('business_id = ?'); binds.push(q.business_id); }
  const rows = await c.env.DB
    .prepare(`SELECT * FROM grant_applications ${where.length ? `WHERE ${where.join(' AND ')}` : ''} ORDER BY created_at DESC LIMIT ? OFFSET ?`)
    .bind(...binds, q.page_size, (q.page - 1) * q.page_size)
    .all<GrantApplicationRow>();
  return c.json((rows.results ?? []).map(serializeGrant));
});

// =py routes/grants.create_grant
router.post('/', requireAuth, requireAdmin, jsonBody(createBody), async (c) => {
  const body = c.req.valid('json');
  const db = c.env.DB;
  const biz = await db.prepare('SELECT id FROM businesses WHERE id = ?').bind(body.business_id).first();
  if (!biz) return c.json({ detail: 'Business not found' }, 404);

  const id = generateId();
  await db.prepare(
    `INSERT INTO grant_applications (id, business_id, status, total_project_cost, acquisition_cost, project_description)
     VALUES (?, ?, 'eligibility_assessed', ?, ?, ?)`
  ).bind(id, body.business_id, body.total_project_cost ?? null, body.acquisition_cost ?? null, body.project_description ?? null).run();
  console.log('grant_application_created', { grant_id: id });
  return c.json(serializeGrant((await loadGrant(db, id))!), 201);
});

// =py routes/grants.update_grant
router.patch('/:grant_id', requireAuth, requireAdmin, jsonBody(updateBody), async (c) => {
  const id = c.req.param('grant_id');
  const db = c.env.DB;
  if (!(await loadGrant(db, id))) return c.json({ detail: 'Grant application not found' }, 404);

  const body = c.req.valid('json');
  const sets: string[] = [];
  const binds: unknown[] = [];
  for (const key of UPDATE_FIELDS) {
    const value = body[key];
    if (value !== undefined) { sets.push(`${key} = ?`); binds.push(typeof value === 'boolean' ? (value ? 1 : 0) : value); }
  }
  if (sets.length) {
    sets.push('updated_at = ?'); binds.push(nowIso());
    await db.prepare(`UPDATE grant_applications SET ${sets.join(', ')} WHERE id = ?`).bind(...binds, id).run();
  }
  return c.json(serializeGrant((await loadGrant(db, id))!));
});

// =py routes/grants.transition_grant_stage
router.patch('/:grant_id/stage', requireAuth, requireAdmin, jsonBody(z.object({ new_stage: z.string() })), async (c) => {
  const id = c.req.param('grant_id');
  const { new_stage } = c.req.valid('json');
  const db = c.env.DB;
  const grant = await loadGrant(db, id);
  if (!grant) return c.json({ detail: 'Grant application not found' }, 404);
  if (!(NOF_STAGES as readonly string[]).includes(new_stage)) return c.json({ detail: `Invalid stage: ${new_stage}` }, 400);

  const allowed = VALID_NOF_TRANSITIONS[grant.status];
  if (!allowed.includes(new_stage as NofStage)) {
    return c.json({ detail: `Cannot transition from ${grant.status} to ${new_stage}. Allowed: [${allowed.map((s) => `'${s}'`).join(', ')}]` }, 422);
  }
  await db.prepare('UPDATE grant_applications SET status = ?, updated_at = ? WHERE id = ?').bind(new_stage, nowIso(), id).run();
  return c.json({ status: 'ok', grant_id: id, new_stage });
});

// =py routes/grants.list_grant_documents
router.get('/:grant_id/documents', requireAuth, async (c) => {
  const id = c.req.param('grant_id')!;
  const db = c.env.DB;
  if (!(await loadGrant(db, id))) return c.json({ detail: 'Grant application not found' }, 404);
  const rows = await db.prepare('SELECT * FROM grant_documents WHERE grant_application_id = ?').bind(id).all<GrantDocumentRow>();
  return c.json((rows.results ?? []).map(serializeDoc));
});

// =py routes/grants.update_grant_document
router.patch('/:grant_id/documents/:doc_id', requireAuth, requireAdmin, jsonBody(docUpdateBody), async (c) => {
  const grantId = c.req.param('grant_id');
  const docId = c.req.param('doc_id');
  const db = c.env.DB;
  const doc = await db.prepare('SELECT * FROM grant_documents WHERE id = ? AND grant_application_id = ?').bind(docId, grantId).first<GrantDocumentRow>();
  if (!doc) return c.json({ detail: 'Grant document not found' }, 404);

  const body = c.req.valid('json');
  const sets: string[] = [];
  const binds: unknown[] = [];
  for (const key of ['status', 'notes', 'received_date', 'reviewed_date'] as const) {
    if (body[key] !== undefined) { sets.push(`${key} = ?`); binds.push(body[key]); }
  }
  if (sets.length) {
    sets.push('updated_at = ?'); binds.push(nowIso());
    await db.prepare(`UPDATE grant_documents SET ${sets.join(', ')} WHERE id = ?`).bind(...binds, docId).run();
  }
  const updated = await db.prepare('SELECT * FROM grant_documents WHERE id = ?').bind(docId).first<GrantDocumentRow>();
  return c.json(serializeDoc(updated!));
});

export default router;
